### DISCUSSION TITLE
Push Notifications API Design

### DISCUSSION THREAD

**Discussion by proneon267** [beeware/toga#1890] [Ideas]
So, here is my preliminary API for push notification widget:
```
Toga.App.send_push_notification(title, text, timer, on_press)
```
It is very simple and doesn't give much customization options. This because I am not sure whether all the platforms do support these customization. Like for example, in macos, I have read that it doesn't allow you to change the notification icon. But other platforms do allow it. So the question arises, can we provide these customization options uniformly? 

Further, it also doesn't check whether the notification sent was successful or not. There maybe cases where the notification will not be shown like in Do not distrub mode is on, or notifications are disabled. But, in each of these cases these settings are set by the user according to their preferences.

If we allow the app developers to know if the notification was sent successfully or if the app has the permission to sent notifications, then I fear these options would be misused. The situation would turn like the situation of websites and adblockers where the websites refuse to work if the notification permission is not given. What if the apps do something similar?

Notifications should act as supplementary to an app. I mean they should not solely depend on the notification to display progress or related things.

Anyways, can you give names of the API that you would want to be implemented?

**Comment by freakboy3742** [comment]
I haven't got any ideas for API names - because I haven't done any research on the topic. However ... it appears that you haven't either (something that you say explicitly - "I'm not sure if all platform support this customisation")

Without even leaving this project, there is an open ticket that contains at some research and design suggestions: #907.

That discussion also addresses one of the biggest issues with what you've described here - the name of what you're suggesting. You've called this a "push notification". A push notification is server initiated. The server "pushes" a notification. If Toga had an API to send a push notification, the user wouldn't see anything (unless they were pushing a notification that was addressed to themselves somehow). As with that older ticket, I suspect what you actually mean is a popup notification, sometimes called a toast (because it pops up like cooked toast from a toaster). 

The *first* step on any new feature proposal is to do platform research. That's why I pointed you at Dan's blog post and video - the first step was to research the capabilities of the new widget he wanted to implement. We're not going to commit to an API design until we have a reasonable degree of confidence that it can be implemented on every platform that Toga supports.

**Reply by proneon267** [reply]
I agree with you. I'll check what I can find and keep you updated.

**Comment by proneon267** [comment]
# Nomenclature
Researching on the topic, I found distinction between the different notifications:
* **Push Notification**: Web server pushed notification.
* **Toast Notification**: In-app notification:
<img src="https://user-images.githubusercontent.com/45512885/235365287-d301a455-16c1-4ae0-8ddf-ad92913d12aa.png" width=50%>

* **User Facing Non in-app notification**:

On Windows:
<img src="https://user-images.githubusercontent.com/45512885/235364828-d385fd5e-fc7a-4a8a-a9af-433ad3104c7f.png" width=50%>
On GTK:
<img src="https://user-images.githubusercontent.com/45512885/235364995-ef28fbd4-823d-4dbf-8f39-448a46fa5496.png" width=50%>
On MacOS:
<img src="https://user-images.githubusercontent.com/45512885/235365185-18ab46be-624f-4da3-9623-55ee2822e994.png" width=50%>

This API design focuses on **User Facing Non in-app notifications**

Windows seems to use the nomenclature of "toast" notification. But, gtk differentiates between in-app(toast) notification and desktop notfication. 

Calling them desktop notification would be good but on mobile platforms, it would not be very logical sounding. Hence, we can call them app notifications would be better.

# The major components of a desktop notfication seems to be:
* App Icon
* App Name
* Title
* Text Body
* Notification Icon
* Buttons(Apple HIG specifies maximum of 4 buttons are allowed.) or other widgets
* Urgency of the notification
* Permission to send notification
* Timeout to show notification
* "on_press" like event which is triggered when the body of the notification is clicked(not the notification buttons).

All platforms have a notification center to display all the notifications from different apps. 

On platforms like Apple, Notification Icon is not changeable. Timeout is also not changeable on some of the platforms. App icon and App Name will remain constant and cannot & should not be changed.

Urgency of the notification determines whether a sound will be played or should the notification be silently displayed in the notification center.

**Reply by freakboy3742** [reply]
That seems like a reasonable summary - it would be good to confirm that these properties are also true of iOS and Android notifications, but intuitively it seems broadly like they would be. 

Regarding the name - I'm not sure a "app" vs "desktop" notification is a distinction worth making; calling these popups "Notifications" seems more than sufficient.

The issue with icons doesn't strike me as a big problem - we can either (a) allow an icon to be specified, but document that that specification might be ignored, or (b) not provide an icon specification, and always provide the app icon where manual specification is required. 

Inconsistencies in the interpretation of timeout is in a similar category - we can establish some basic categories (e.g., "won't disappear", "short lived, "long lived") and then provide a "best effort" interpretation of those categories on each platform. That won't exposed the full flexibility that some platforms allow, but it will functionally allow for most use cases.

"Urgency" seems a bit more vague - there might be a similar mapping here, but it's not immediately obvious to me what that would be.

The only other item that seems underspecified is the "buttons" aspect - you mention the Apple HIG allows for up to 4 buttons, but I have to assume there are similar restrictions on other platforms. I'd almost be inclined to say "iteration 1" of this design only allows for "on_press" style event handling, and defer "additional actions" as something we look at in a second design iteration.

**Reply by proneon267** [reply]
Notification urgency can be of:
* Low - show the notification without any sound and the notification timeout should be finite
* Normal - show the notification with a sound and the notification timeout should be finite
* Critical - show the notification with a sound and the notification should not timeout automatically and should only expire after the user dismisses the notification themselves. 

Also, what about notification permission? Apple's user notification and mobile platform require prior permission to send notification. But, windows doesn't enforce such permissions.

**Reply by freakboy3742** [reply]
So - you're proposing an interpretation of "urgency" that seems to overlap with 2 other properties - "makes a sound", and "timeout". Is Urgency a fundamental property of notifications, or something else entirely? What are the raw platform interpretations of "urgency"? 

As for permissions - you're the one proposing the designs, so you tell me what is needed to handle permissions :-) I have no idea (without doing research) what needs to be done to support the required permissions on macOS. However, it seems highly likely that this will require some sort of solution for beeware/briefcase#547.

**Reply by proneon267** [reply]
# Urgency/Priority
# On Windows:
![image](https://user-images.githubusercontent.com/45512885/235405329-8f653e6a-f81d-4eb2-b442-eb5491578885.png)
![image](https://user-images.githubusercontent.com/45512885/235404670-dee9b3c7-2878-46ac-9fdb-c2c6aa85a85d.png)
Here cached mean the notifications are not processed until the user signs in.

# On GTK:
![image](https://user-images.githubusercontent.com/45512885/235405044-004b2f55-cf08-41f8-88dd-5b3e03a7995d.png)

# On Apple Devices:
![image](https://user-images.githubusercontent.com/45512885/235405231-cfbaa617-9c23-44d1-9312-e97b104299e6.png)

As for permissions on apple devices, I found: https://developer.apple.com/documentation/usernotifications/asking_permission_to_use_notifications

**Reply by freakboy3742** [reply]
So - that's raw data from research, but isn't a proposal.  

It's clear that all platforms have a concept of "urgency", but it isn't obvious to me that the levels all inherently (or consistently) match to concepts of "timeout" *or* "sound" - "urgency" seems to map to a range of ideas related to when the notification will interrupt the user, or override battery settings.

So what are you proposing?

**Reply by proneon267** [reply]
I am proposing that we have a Notification urgency option, that will have: `Low`, `Normal`, `Critical` options with each mapped to the behavior of the native platform notification.

I'll write a preliminary API for the features described so far and let you know.

**Reply by proneon267** [reply]
`Toga.App.send_notification(title, text, timer, priority, on_press)`
* `title` - title of the notification.
* `text` - body text of the notification.
* `timer` - `short lived` `won't disapper` `long lived` - Not actual API values, just pseudo values.
* `priority` - `Low` `Normal` `Critical`
* `on_press` - takes the function to call when the body of the notification is clicked.
* `additional_widgets_box` - takes a `Toga.Box` containing the widgets to be displayed on the notfication - Will not be included in the first iteration of the notifcation design.

I think the `timer` and `priority` parameters have overlapping goals. If a notification is `Low` priority, it should be `short lived` and if it is `Critical` then it should be `won't disappear`. 

@freakboy3742 please take a look at this and let me know.

**Reply by freakboy3742** [reply]
> `Toga.App.send_notification(title, text, timer, priority, on_press)`

"Send" implies it's going somewhere; I'd prefer `show_notification()`, since it's a lot more "local" as terminology.
> 
> * `title` - title of the notification.
> * `text` - body text of the notification.
> * `timer` - `short lived` `won't disapper` `long lived` - Not actual API values, just pseudo values.

Timer implies there's a configurable duration. Something like `transient = True/False` or `persistent=False/True` would make more sense to me.

> * `priority` - `Low` `Normal` `Critical`
> * `on_press` - takes the function to call when the body of the notification is clicked.
> * `additional_widgets_box` - takes a `Toga.Box` containing the widgets to be displayed on the notfication - Will not be included in the first iteration of the notifcation design.

Not sure this will be viable as a design, but since it's not in this iteration, I guess it doesn't matter at this stage.

> I think the `timer` and `priority` parameters have overlapping goals. If a notification is `Low` priority, it should be `short lived` and if it is `Critical` then it should be `won't disappear`.

If they're overlapping, why define both? What's the implementation if `priority=Critical, timer=short lived` or `priority=Low, timer=won't disappear`? 
 
> @freakboy3742 please take a look at this and let me know.

There's no need to ping me every time you want a response. Firstly, it implies nobody else is welcome to respond; secondly, I get notified of every post in every BeeWare forum. The only reason to tag me specifically is if (a) there is some reason that my input *specifically* is required, and I haven't been involved in the discussion to date, or (b) you've posted something needing a response and haven't received one for a couple of days (allowing for weekends, etc when someone may not be able to post).

**Reply by proneon267** [reply]
> "Send" implies it's going somewhere; I'd prefer show_notification(), since it's a lot more "local" as terminology.

I agree.
> Timer implies there's a configurable duration. Something like transient = True/False or persistent=False/True would make more sense to me.

I agree.
> If they're overlapping, why define both? What's the implementation if priority=Critical, timer=short lived or priority=Low, timer=won't disappear?

I meant which one of the following: `priority` and `timer` should be keep, since they are overlapping. For example, a `critical` notification should be `won't disappear`. 

> There's no need to ping me every time you want a response. Firstly, it implies nobody else is welcome to respond; secondly, I get notified of every post in every BeeWare forum. The only reason to tag me specifically is if (a) there is some reason that my input specifically is required, and I haven't been involved in the discussion to date, or (b) you've posted something needing a response and haven't received one for a couple of days (allowing for weekends, etc when someone may not be able to post).

Won't happen again :)

**Reply by freakboy3742** [reply]
> > If they're overlapping, why define both? What's the implementation if priority=Critical, timer=short lived or priority=Low, timer=won't disappear?
> 
> I meant which one of the following: `priority` and `timer` should be keep, since they are overlapping. For example, a `critical` notification should be `won't disappear`.

Ah - I didn't read what you'd written as a question. If that's the case, I'd say "priority" is the option that makes more sense - it's a "higher level" concept that is required on the native APIs anyway, and the implied mapping to notification persistence makes sense.

**Comment by NickVanelli** [comment]
friend, I have the same goal, can you give me some form of contact with you so that we can talk about our projects?

**Comment by annunal** [comment]
Is there any news about this subject ? It seems stopping at June 2023;  was a solution found in the meantime ?

**Reply by freakboy3742** [reply]
No, there hasn't been any further work on this proposal since the last update in May. We're still interested in adding it, but it's not on the [short-term roadmap of the core team](https://beeware.org/news/buzz/2024q1-roadmap/).
